File: 238_Product_of_Array_Except_Self/238_1.py
from typing import List

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = [1] * n
        prefix, suffix = 1, 1
        for i in range(n):
            res[i] *= prefix
            res[n-1-i] *= suffix
            prefix *= nums[i]
            suffix *= nums[n-1-i]
        return res

File: 238_Product_of_Array_Except_Self/238_1.c
int* productExceptSelf(int* nums, int numsSize, int* returnSize){
    *returnSize = numsSize;
    int *res = (int *)malloc(sizeof(int) * numsSize);
    for (int i = 0; i < numsSize; i++) {
        res[i] = 1;
    }
    int left = 1, right = 1;
    for (int i = 0; i < numsSize; i++) {
        res[i] *= left;
        left *= nums[i];
    }
    for (int i = numsSize - 1; i >= 0; i--) {
        res[i] *= right;
        right *= nums[i];
    }
    return res;
}
